LunrCoreLmdb/SerializeContext.cs

```diff
@@ -0,0 +1,88 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+using Lunr;
+
+namespace LunrCoreLmdb
+{
+    public sealed class SerializeContext
+    {
+        public const int FormatVersion = 1;
+
+        private readonly BinaryWriter _bw;
+
+        public SerializeContext(BinaryWriter bw, int version = FormatVersion)
+        {
+            _bw = bw;
+            if (Version > FormatVersion)
+                throw new Exception("Tried to write an object with a version that is too new");
+            Version = version;
+            _bw.Write(Version);
+        }
+
+        public int Version { get; }
+
+        public void Write(int value) => _bw.Write(value);
+
+        public void Write(double value) => _bw.Write(value);
+
+        public bool Write(bool value)
+        {
+            _bw.Write(value);
+            return value;
+        }
+
+        public void Write(byte[] value)
+        {
+            _bw.Write(value.Length);
+            _bw.Write(value);
+        }
+
+        public void Write(string value)
+        {
+            _bw.Write(value.Length);
+            foreach (var c in value)
+                _bw.Write((byte) c);
+        }
+
+        internal static readonly Dictionary<Type, (Func<object, byte[]>, Func<byte[], object> deserialize)> KnownTypes;
+
+        static SerializeContext()
+        {
+            KnownTypes = new Dictionary<Type, (Func<object, byte[]>, Func<byte[], object> deserialize)>
+            {
+                {typeof(short), (v => BitConverter.GetBytes((short) v), b => BitConverter.ToInt16(b, 0))},
+                {typeof(int), (v => BitConverter.GetBytes((int) v), b => BitConverter.ToInt32(b, 0))},
+                {typeof(long), (v => BitConverter.GetBytes((long) v), b => BitConverter.ToInt64(b, 0))},
+                {typeof(ushort), (v => BitConverter.GetBytes((ushort) v), b => BitConverter.ToUInt16(b, 0))},
+                {typeof(uint), (v => BitConverter.GetBytes((uint) v), b => BitConverter.ToUInt32(b, 0))},
+                {typeof(ulong), (v => BitConverter.GetBytes((ulong) v), b => BitConverter.ToUInt64(b, 0))},
+                {typeof(float), (v => BitConverter.GetBytes((int) v), b => BitConverter.ToSingle(b, 0))},
+                {typeof(double), (v => BitConverter.GetBytes((int) v), b => BitConverter.ToDouble(b, 0))},
+                {typeof(bool), (v => BitConverter.GetBytes((bool) v), b => BitConverter.ToBoolean(b, 0))},
+                {typeof(char), (v => BitConverter.GetBytes((char) v), b => BitConverter.ToChar(b, 0))}
+            };
+
+            KnownTypes.Add(typeof(string), (v => Encoding.UTF8.GetBytes((string) v), b => Encoding.UTF8.GetString(b)));
+            KnownTypes.Add(typeof(Slice), (v =>
+            {
+                var (start, length) = (Slice) v;
+                return Encoding.UTF8.GetBytes($"{start}/{length}");
```

Comment:
Using `start + "/" + length` should be considerably better here.

Reply:
``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18363.1256 (1909/November2018Update/19H2)
Intel Core i9-9820X CPU 3.30GHz, 1 CPU, 20 logical and 10 physical cores
.NET Core SDK=5.0.101
  [Host]     : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
  Job-TLTHGY : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT

RunStrategy=Throughput  

```
|      Method |      Mean |    Error |   StdDev |
|------------ |----------:|---------:|---------:|
| Interpolate | 130.67 ns | 0.322 ns | 0.301 ns |
|         Add |  70.81 ns | 0.151 ns | 0.126 ns |

Reply:
Correct! Time to disabuse myself of the prettiness of interpolation.